### ISSUE TITLE
Unable to learn simple catch game

### ISSUE THREAD

**Issue by hmate9** [matthiasplappert/keras-rl#51]
I've made custom environment, where the fruit is falling and you control a paddle to catch it: https://github.com/hmate9/gym-catch/blob/master/gym_catch/envs/catch_env.py

I've tried to use keras-rl to reimplement this: https://gist.github.com/EderSantana/c7222daa328f0e885093

The same game, catching a fruit, and their implementation finds a good model in a couple of minutes which catches nearly 100% of the time. 

Here is the code for learning with keras-rl that I wrote: https://gist.github.com/hmate9/49758ee1117ae55616f45d72186834a5

The code with keras-rl does not converge, and does not even seem to be better than random even after running for hours. 

Does anyone know why this is? Did I write the environment wrong or am I using keras-rl wrong?

Your answer is greatly appreciated, I have not been able to solve this for a day now.

**Comment by hmate9** [comment]
I've also tried
`policy = EpsGreedQPolicy(0.1)` but that did not work either.

@rmst I noticed you were able to help in another thread to someone with a similar problem. Do you know what's wrong here?

**Comment by matthiasplappert** [comment]
Try to increase the warm-up phase a bit, e.g. 1000, and decrease the target model updates, e.g. 1000 as well. The learning rate also seems very high, I'd try something like 1e-3. Also make sure that your env us properly implemented, if you didn't already.

**Comment by hmate9** [comment]
Thanks for your response @matthiasplappert 

I've tried your suggestions, but no change in the results. Even after 20,000 episodes, the performance is only as good as random chance. 

Also, I am sure my environment is implemented correctly. It is an extremely simple env, only 50 lines (when not looking at the render function)

**Comment by matthiasplappert** [comment]
Can you please post the output that you get when training your agent?

**Comment by hmate9** [comment]
Here is snippet after 3700 episodes:

` 18080/100000: episode: 3691, duration: 0.023s, episode steps: 4, steps per second: 173, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 0.750 [0.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.041399, mean_absolute_error: 0.495927, mean_q: -0.566884
 18084/100000: episode: 3692, duration: 0.025s, episode steps: 4, steps per second: 158, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 0.750 [0.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.045398, mean_absolute_error: 0.476438, mean_q: -0.510456
 18088/100000: episode: 3693, duration: 0.018s, episode steps: 4, steps per second: 217, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 0.750 [0.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.033347, mean_absolute_error: 0.508438, mean_q: -0.598909
 18092/100000: episode: 3694, duration: 0.013s, episode steps: 4, steps per second: 306, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 1.500 [1.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.044392, mean_absolute_error: 0.498428, mean_q: -0.571252
 18096/100000: episode: 3695, duration: 0.017s, episode steps: 4, steps per second: 231, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 1.500 [1.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.050968, mean_absolute_error: 0.508118, mean_q: -0.602726
 18100/100000: episode: 3696, duration: 0.021s, episode steps: 4, steps per second: 193, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 1.500 [1.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.049489, mean_absolute_error: 0.487274, mean_q: -0.532419
 18104/100000: episode: 3697, duration: 0.015s, episode steps: 4, steps per second: 267, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 1.500 [1.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.053799, mean_absolute_error: 0.465268, mean_q: -0.506819
 18108/100000: episode: 3698, duration: 0.017s, episode steps: 4, steps per second: 231, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 0.750 [0.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.037638, mean_absolute_error: 0.462787, mean_q: -0.515579
 18112/100000: episode: 3699, duration: 0.020s, episode steps: 4, steps per second: 200, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 0.750 [0.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.046121, mean_absolute_error: 0.464610, mean_q: -0.514178
 18120/100000: episode: 3700, duration: 0.030s, episode steps: 8, steps per second: 265, episode reward: 0.000, mean reward: 0.000 [-1.000, 1.000], mean action: 1.000 [0.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.048673, mean_absolute_error: 0.473400, mean_q: -0.523688
 18124/100000: episode: 3701, duration: 0.022s, episode steps: 4, steps per second: 180, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 0.750 [0.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.041581, mean_absolute_error: 0.489766, mean_q: -0.567378
 18128/100000: episode: 3702, duration: 0.016s, episode steps: 4, steps per second: 258, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 0.750 [0.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.042322, mean_absolute_error: 0.514674, mean_q: -0.622504
 18132/100000: episode: 3703, duration: 0.017s, episode steps: 4, steps per second: 233, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 0.750 [0.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.055914, mean_absolute_error: 0.487348, mean_q: -0.547432
 18136/100000: episode: 3704, duration: 0.016s, episode steps: 4, steps per second: 258, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 1.500 [1.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.046242, mean_absolute_error: 0.475554, mean_q: -0.536386
 18140/100000: episode: 3705, duration: 0.020s, episode steps: 4, steps per second: 197, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 0.750 [0.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.047849, mean_absolute_error: 0.483586, mean_q: -0.541547
 18144/100000: episode: 3706, duration: 0.016s, episode steps: 4, steps per second: 247, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 1.500 [1.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.049242, mean_absolute_error: 0.472304, mean_q: -0.521400
 18148/100000: episode: 3707, duration: 0.021s, episode steps: 4, steps per second: 191, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 1.000 [0.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.041994, mean_absolute_error: 0.473593, mean_q: -0.527861
 18152/100000: episode: 3708, duration: 0.016s, episode steps: 4, steps per second: 251, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 1.000 [0.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.060627, mean_absolute_error: 0.490083, mean_q: -0.547527
 18156/100000: episode: 3709, duration: 0.013s, episode steps: 4, steps per second: 304, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 1.500 [1.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.050812, mean_absolute_error: 0.491437, mean_q: -0.550912
 18160/100000: episode: 3710, duration: 0.019s, episode steps: 4, steps per second: 212, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 1.500 [1.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.046626, mean_absolute_error: 0.477001, mean_q: -0.519709
 18164/100000: episode: 3711, duration: 0.014s, episode steps: 4, steps per second: 293, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 0.750 [0.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.046759, mean_absolute_error: 0.489565, mean_q: -0.565330
 18168/100000: episode: 3712, duration: 0.013s, episode steps: 4, steps per second: 306, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 1.500 [1.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.033925, mean_absolute_error: 0.489299, mean_q: -0.550303
 18172/100000: episode: 3713, duration: 0.017s, episode steps: 4, steps per second: 240, episode reward: -1.000, mean reward: -0.250 [-1.000, 0.000], mean action: 0.750 [0.000, 2.000], mean observation: 0.080 [0.000, 1.000], loss: 0.058524, mean_absolute_error: 0.502469, mean_q: -0.555000`

It is like this throughout. 

The network description:

`____________________________________________________________________________________________________
Layer (type)                     Output Shape          Param #     Connected to                     
====================================================================================================
flatten_1 (Flatten)              (None, 25)            0           flatten_input_1[0][0]            
____________________________________________________________________________________________________
activation_1 (Activation)        (None, 25)            0           flatten_1[0][0]                  
____________________________________________________________________________________________________
dense_1 (Dense)                  (None, 100)           2600        activation_1[0][0]               
____________________________________________________________________________________________________
activation_2 (Activation)        (None, 100)           0           dense_1[0][0]                    
____________________________________________________________________________________________________
dense_2 (Dense)                  (None, 100)           10100       activation_2[0][0]               
____________________________________________________________________________________________________
activation_3 (Activation)        (None, 100)           0           dense_2[0][0]                    
____________________________________________________________________________________________________
dense_3 (Dense)                  (None, 3)             303         activation_3[0][0]               
____________________________________________________________________________________________________
activation_4 (Activation)        (None, 3)             0           dense_3[0][0]                    
====================================================================================================
Total params: 13003
____________________________________________________________________________________________________`

**Comment by hmate9** [comment]
Here it is with verbose=1:

Interval 1 (0 steps performed)
1000/1000 [==============================] - 0s - reward: -0.0860          
168 episodes - episode_reward: -0.512 [-1.000, 4.000]

Interval 2 (1000 steps performed)
1000/1000 [==============================] - 4s - reward: -0.1700          
210 episodes - episode_reward: -0.810 [-1.000, 2.000] - loss: 0.191 - mean_absolute_error: 0.249 - mean_q: 0.066

Interval 3 (2000 steps performed)
1000/1000 [==============================] - 3s - reward: -0.1340        
192 episodes - episode_reward: -0.703 [-1.000, 2.000] - loss: 0.188 - mean_absolute_error: 0.266 - mean_q: -0.012

Interval 4 (3000 steps performed)
1000/1000 [==============================] - 3s - reward: -0.1360        
193 episodes - episode_reward: -0.699 [-1.000, 2.000] - loss: 0.174 - mean_absolute_error: 0.263 - mean_q: -0.098

Interval 5 (4000 steps performed)
1000/1000 [==============================] - 3s - reward: -0.1620        
206 episodes - episode_reward: -0.786 [-1.000, 2.000] - loss: 0.181 - mean_absolute_error: 0.313 - mean_q: 0.007

Interval 6 (5000 steps performed)
1000/1000 [==============================] - 3s - reward: -0.1480        
199 episodes - episode_reward: -0.744 [-1.000, 3.000] - loss: 0.179 - mean_absolute_error: 0.349 - mean_q: -0.138

Interval 7 (6000 steps performed)
1000/1000 [==============================] - 3s - reward: -0.1460        
198 episodes - episode_reward: -0.737 [-1.000, 2.000] - loss: 0.185 - mean_absolute_error: 0.355 - mean_q: -0.067

Interval 8 (7000 steps performed)
1000/1000 [==============================] - 3s - reward: -0.1400        
195 episodes - episode_reward: -0.718 [-1.000, 2.000] - loss: 0.152 - mean_absolute_error: 0.360 - mean_q: -0.163

Interval 9 (8000 steps performed)
1000/1000 [==============================] - 3s - reward: -0.1680        
209 episodes - episode_reward: -0.804 [-1.000, 3.000] - loss: 0.155 - mean_absolute_error: 0.365 - mean_q: -0.207

**Comment by matthiasplappert** [comment]
Is that output with Boltzmann or eps-greedy? I'd recommend eps-greedy since Boltzmann can be tricky. Maybe try to reduce eps to 0.01 in this setting. I can also take a more detailed look at your code tomorrow if that doesn't help either.

**Comment by hmate9** [comment]
That was with eps set at 0.1

I have tried it with 0.01 just now, same output in terms of rewards and it not converging at all.

Thanks @matthiasplappert, I'd really appreciate it if you could look over the code!

**Comment by hmate9** [comment]
Should be noted that this is not supposed to be a difficult environment to solve. Someone else's code (which does not use keras-rl) converges after a few seconds.

**Comment by matthiasplappert** [comment]
Yeah, that really shouldn't be a problem which is why I suspect that there's a subtile bug somewhere in the environment or the hyperparameter setup. I'll have another look tomorrow.

**Comment by hmate9** [comment]
I've found the problem.

In my environment implementation of the `step` function when I returned the state I returned a reference to the current state, and not a deep copy. This messed up things completely of course.

Thanks @matthiasplappert for your time!

**Comment by matthiasplappert** [comment]
Ah yes, that explains it. Good catch. I also consider this a keras-rl bug since you would really expect it to do the right thing here. I've therefore fixed this problem in https://github.com/matthiasplappert/keras-rl/commit/5fb1255fd74adb6e5eefb237890cdd2d30cb3365. Thanks for reporting this and figuring out the problem!

Also, completely unrelated: Would it be okay if I use your catch env code in an example that demonstrates how to implement a custom environment? I'd of course include a comment that refers to you and your Github repo as the author.

**Comment by hmate9** [comment]
@matthiasplappert Absolutely, feel free to do it. I'm glad you can find a use for it.
